### ISSUE TITLE
ASD confidence/score

### ISSUE THREAD

**Issue by dberghi** [TaoRuijie/TalkNet_ASD#37]
Hi there,

Thank you for releasing your code and models, very impressive results! :) 

I am evaluating your TalkNet demo model on a new dataset. I now need to draw a precision-recall curve, so I was expecting to get something like a per-frame ASD confidence value in the range [0; 1] so that I can gradually vary the confidence threshold from 0% to 100% and compute a pair of precision and recall rates at each step.
Instead, TalkNet provides per-frame scores that seem to range from about -2 or -3 to +2 or +3, more or less, without clear max or min boundaries, where positive values are active and negative silent frames.

From my understanding, the last FC layer is followed by a softmax operation so shouldn't the output be expressed in terms of [0; 1] confidence? 
Is there a way to convert the output score into a confidence? I was thinking of simply applying a Sigmoid function to the output score, but perhaps I am missing something.

Thank you again for your work, looking forward to hearing back from you!
Davide

**Comment by TaoRuijie** [comment]
That is an interesting problem. I am not very sure the best solution. Just share my idea. During evaluation, I use the output from the FC directly as the score (please check the code in loss_av.py). So the output size is like [Batch_size, video_length, 2], 2 means [score_0, score_1], for 'non-speaking' and 'speaking'. So I use score_1 and put it on the demo video.

So I think you can add the softmax, that might be a good choice to make the score_0 and score_1 within 0 to 1, Then you can use the score_1 to compute the curve you want.

Ruijie

**Comment by dberghi** [comment]
Interesting! 
Thank you for the suggestion, I have just applied it. I have also removed the round operation from line 251 
allScore = numpy.round((numpy.mean(numpy.array(allScore), axis = 0)), 1).astype(float) to have more confidence steps (otherwise it only returns 0, 0.1, 0.2, 0.3, ..., 1).

Probably, the Sigmoid function would have also worked fine as it maps the scores in the range 0 to 1. However, it presents the problem of choosing a suitable slope. So, I like your solution, thanks! :)

For whoever needs to express the ASD in terms of 0 to 1 confidence in the TalkNet demo, here are listed the key steps:
1) Add softmax operation in lossAV in the loss.py script (after line 14 add x = F.softmax(x, dim=-1))
2) Remove round operation from line 251 in demoTalkNet.py
(optional for visualization)
3) Modify line 274 with clr = colorDict[int((face['score'] >= 0.5))] to use 50% confidence threshold.

Hope this might be useful to someone else in the future too.

Davide
